db-engine/src/main/java/com/splicemachine/db/iapi/sql/dictionary/DatabaseDescriptor.java

```diff
@@ -265,4 +272,50 @@ public void writeExternal(ObjectOutput output) throws IOException {
     public void setDataDictionary(DataDictionary dataDictionary) {
         this.dataDictionary = dataDictionary;
     }
+
+    /**
+     * Drop this database.
+     * Drops the database if it is empty.
+     * @throws StandardException Schema could not be dropped.
+     */
+    public void drop(LanguageConnectionContext lcc,
+                     Activation activation) throws StandardException
+    {
+        DataDictionary dd = getDataDictionary();
+        DependencyManager dm = dd.getDependencyManager();
+        TransactionController tc = lcc.getTransactionExecute();
+
+        // First drop restrict the default SPLICE schema:
+        SchemaDescriptor spliceSchemaDesc = dd.getSchemaDescriptor(getUUID(), Property.DEFAULT_USER_NAME, tc, false);
+        if (spliceSchemaDesc != null) {
+            spliceSchemaDesc.drop(lcc, activation);
+        }
+
+        /*
+         ** Make sure the database is empty.
+         ** In the future we want to drop everything
+         ** in the database if it is CASCADE.
+         */
+        if (!dd.isDatabaseEmpty(this))
+        {
+            throw StandardException.newException(SQLState.LANG_DATABASE_NOT_EMPTY, getDatabaseName());
+        }
+
+        /* Prepare all dependents to invalidate.  (This is there chance
+         * to say that they can't be invalidated.  For example, an open
+         * cursor referencing a table/view that the user is attempting to
+         * drop.) If no one objects, then invalidate any dependent objects.
+         */
+        dm.invalidateFor(this, DependencyManager.DROP_DATABASE, lcc);
+
+        dd.dropDatabaseDescriptor(getDatabaseName(), tc);
+
+        /*
+         ** If we have dropped the current default databae,
+         ** then we will set the default to null.  The
+         ** LCC is free to set the new default database to
+         ** some system defined default.
+         */
+        // lcc.resetSchemaUsages(activation, getDatabaseName()); XXX (arnaud multidb) implement that?
```

Comment:
It is not clear to me what behavior we should support when the default database in the connection is dropped. Right now, after dropping the default database in a connection, we still allow new schema to be created, but using the database id that has been dropped, thus creating dangling schemas:
```
splice> create database test_db;
0 rows inserted/updated/deleted
ELAPSED TIME = 18 milliseconds
splice> connect 'jdbc:splice://localhost:1527/test_db;user=splice;password=admin' as testdb_con;
ELAPSED TIME = 18 milliseconds
splice> show schemas;
TABLE_SCHEM                   
------------------------------
SPLICE                        
SYS                           
SYSFUN                        
SYSIBM                        
SYSIBMADM                     
SYSVW                         

6 rows selected
ELAPSED TIME = 19 milliseconds
splice> set connection splice;
ELAPSED TIME = 0 milliseconds
splice> select * from sys.sysdatabases;
DATABASEID                          |DATABASENAME                                                                                                                    |AUTHORIZATIONID                                                                                                                 
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
ecec0361-0175-66e7-a8b7-000001040c80|TEST_DB                                                                                                                         |PLACEHOLDER                                                                                                                     
a816c00e-0175-66e7-a8b7-000001040c80|SPLICEDB                                                                                                                        |PLACEHOLDER                                                                                                                     

2 rows selected
ELAPSED TIME = 8 milliseconds
splice> drop database test_db restrict;
0 rows inserted/updated/deleted
ELAPSED TIME = 46 milliseconds
splice> set connection testdb_con;
ELAPSED TIME = 0 milliseconds
splice> create schema test_sch;
0 rows inserted/updated/deleted
ELAPSED TIME = 29 milliseconds
splice> show schemas;
TABLE_SCHEM                   
------------------------------
SYS                           
SYSFUN                        
SYSIBM                        
SYSIBMADM                     
SYSVW                         

5 rows selected
ELAPSED TIME = 15 milliseconds
splice> select * from sys.sysschemas;
SCHEMAID                            |SCHEMANAME                                                                                                                      |AUTHORIZATIONID                                                                                                                 |DATABASEID                          
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
2a534337-0175-66e7-a8b7-000001040c80|TEMP                                                                                                                            |SPLICE                                                                                                                          |a816c00e-0175-66e7-a8b7-000001040c80
bde2434f-0175-66e7-a8b7-000001040c80|TEMP2                                                                                                                           |SPLICE                                                                                                                          |2cd98344-0175-66e7-a8b7-000001040c80
d8d6436f-0175-66e7-a8b7-000001040c80|TEST_SCH                                                                                                                        |SPLICE                                                                                                                          |ecec0361-0175-66e7-a8b7-000001040c80
c013800d-00f8-5b53-28a9-00000019ed88|SYSIBM                                                                                                                          |SPLICE                                                                                                                          |a816c00e-0175-66e7-a8b7-000001040c80
62bfdcdb-ea53-4339-bd1b-018e411b8852|SYSIBMADM                                                                                                                       |SPLICE                                                                                                                          |a816c00e-0175-66e7-a8b7-000001040c80
8000000d-00d0-fd77-3ed8-000a0a0b1900|SYS                                                                                                                             |SPLICE                                                                                                                          |a816c00e-0175-66e7-a8b7-000001040c80
c013800d-00fb-2641-07ec-000000134f30|SYSCAT                                                                                                                          |SPLICE                                                                                                                          |a816c00e-0175-66e7-a8b7-000001040c80
c013800d-00fb-2642-07ec-000000134f30|SYSFUN                                                                                                                          |SPLICE                                                                                                                          |a816c00e-0175-66e7-a8b7-000001040c80
c013800d-00fb-2643-07ec-000000134f30|SYSPROC                                                                                                                         |SPLICE                                                                                                                          |a816c00e-0175-66e7-a8b7-000001040c80
c013800d-00fb-2644-07ec-000000134f30|SYSSTAT                                                                                                                         |SPLICE                                                                                                                          |a816c00e-0175-66e7-a8b7-000001040c80
c013800d-00fb-2647-07ec-000000134f30|NULLID                                                                                                                          |SPLICE                                                                                                                          |a816c00e-0175-66e7-a8b7-000001040c80
c013800d-00fb-2648-07ec-000000134f30|SQLJ                                                                                                                            |SPLICE                                                                                                                          |a816c00e-0175-66e7-a8b7-000001040c80
c013800d-00fb-2646-07ec-000000134f30|SYSCS_DIAG                                                                                                                      |SPLICE                                                                                                                          |a816c00e-0175-66e7-a8b7-000001040c80
c013800d-00fb-2649-07ec-000000134f30|SYSCS_UTIL                                                                                                                      |SPLICE                                                                                                                          |a816c00e-0175-66e7-a8b7-000001040c80
c013800d-00fb-264d-07ec-000000134f30|SYSVW                                                                                                                           |SPLICE                                                                                                                          |a816c00e-0175-66e7-a8b7-000001040c80
80000000-00d2-b38f-4cda-000a0a412c00|SPLICE                                                                                                                          |SPLICE                                                                                                                          |a816c00e-0175-66e7-a8b7-000001040c80

16 rows selected
ELAPSED TIME = 28 milliseconds
```
As we can see, the schema `test_sch` 's database id is still pointing to `ecec0361-0175-66e7-a8b7-000001040c80`, which is the database id of `test_db`.

Reply:
We should check the existence of the database before creating the schema, so that dangling schemas can be avoided.

Reply:
That's a very good point, thanks!
Is it alright with you if I jira that for later?

Reply:
A JIRA to track is fine. Thanks!

Reply:
Tracked by DB-10591